File: assiist_back_end/db/repositories/interfaces/note_repository.py
from abc import ABC, abstractmethod
from typing import List, Optional
import uuid

from assiist_back_end.models.note import Note

class NoteRepository(ABC):
    """Interface for data operations related to Notes."""

    @abstractmethod
    async def add(self, user_id: str, contact_id: str, note: Note) -> Note:
        """Adds a new note to a contact's subcollection."""
        pass

    @abstractmethod
    async def get_by_id(self, user_id: str, contact_id: str, note_id: str) -> Optional[Note]:
        """Gets a specific note by its ID."""
        pass

    @abstractmethod
    async def get_for_contact(self, user_id: str, contact_id: str) -> List[Note]:
        """Gets all notes for a specific contact."""
        pass

    @abstractmethod
    async def update(self, user_id: str, contact_id: str, note_id: str, update_data: dict) -> Optional[Note]:
        """Updates an existing note."""
        pass

    @abstractmethod
    async def delete(self, user_id: str, contact_id: str, note_id: str) -> bool:
        """Deletes a note."""
        pass 

File: assiist_front_end/lib/core/repositories/note_repository.dart
import 'package:assiist_front_end/core/models/note.dart';

abstract class NoteRepository {
  /// Creates a new note associated with a contact.
  Future<Note> createNote(Note note, String contactId);

  /// Retrieves all notes associated with a specific contact.
  Future<List<Note>> getNotesForContact(String contactId);

  /// Retrieves a specific note by its ID.
  Future<Note?> getNoteById(String contactId, String noteId);

  // Potential future methods:
  // Future<Note?> getNoteById(String noteId);
  // Future<Note?> updateNote(String noteId, Map<String, dynamic> updateData);
  // Future<bool> deleteNote(String noteId);
}
